WordPress/gutenberg
# Navigation: add unit tests for onChange handler and fix cases around custom links tags and post-format

Fixes https://github.com/WordPress/gutenberg/issues/31254 https://github.com/WordPress/gutenberg/issues/31379

https://user-images.githubusercontent.com/1270189/116925012-802e4380-ac0d-11eb-84d1-58d4b9e8f0d5.mp4

### Changes of interest:

* **Pulls out the LinkControl onChange function** in Navigation Link so we can add **unit test cases**
* **Custom links are now persisted with "kind:custom"** instead of serializing misc link control specific items, like (tel, URL, internal mailto).
* **Consistently save as type:tag**. We were inconsistently saving "type:tag" as a tag variation placeholder and "type:post_tag" when a full tag link was inserted. I'm not sure of the original reasons for favoring "tag" versus the actual "post_tag" name.
* **Toggling opensInNewTab on an existing link no longer wipes out other block attributes**
* **Updates fetchLinkSuggestions to return a type:post-format and kind:taxonomy**

~~Main feedback questions~~:
* ~~Did we intend to persist "internal/tel/URL/mailto" or is it okay to map these to a plain custom type?~~ A: good to map to custom
* ~~Are post-formats and post type archives similar or different use cases?~~ post-format is a `taxonomy` with a name of `post_format`
* ~~If folks prefer, instead of adding a `getNavigationLinkType` helper I can create a function for the LinkControl onChange function. It'd be a little bit of a refactor, but nice for unit testing, as we can isolate the data transform cases.~~ I think it's worth the unit tests at this point, so I'll isolate and refactor this.

### Testing Instructions

Unit Tests:
- Verify that the new test cases make sense. Run them locally with `npm run test-unit packages/block-library/src/navigation-link/test/edit.js`

Custom Links:
- Visit the post or site editor
- Add a navigation block 
- Add a custom link variation
- Type email link like "mailto:foo@example.com", Type a custom url like "wordpress.org", Type a telephone link like: "tel:1234567", Type an internal link like "#foo"
- Save, and check in code view that the type value for the navigation link is "custom"

Tag:
- Visit the post or site editor
- Add a navigation block 
- Add an item, select tag variation
- Do not click on a tag suggestion. Click out then inspect code view.
- Verify that type = tag
- Select a tag suggestion
- Verify that type = tag in code view

Opens In New Tab
- Visit the post or site editor
- Add a navigation block 
- Add a post link
- Click on the link symbol, then toggle opensInNewTab.
- Verify that id is still set in code view

Post Formats
- A post format link looks like http://wordpress.local/type/quote/. This points at the archive of a post type format like quote or image.
- Install a theme that supports post-formats like Twenty Thirteen
- Create a post using one of the post-format types like "Quote"
- Add a navigation block
- Add a navigation link using the general link variation
- Search for the post-format and individual quote
- Verify that the link to the post-format has a type of "post_format" and a kind of "taxonomy"
